internal/runtimehandlerhooks/high_performance_hooks.go

```diff
@@ -124,3 +183,71 @@ func setCPUSLoadBalancing(c *oci.Container, enable bool, schedDomainDir string)
 
 	return nil
 }
+
+func setIRQLoadBalancing(c *oci.Container, enable bool) error {
+	if c.Spec().Linux == nil ||
+		c.Spec().Linux.Resources == nil ||
+		c.Spec().Linux.Resources.CPU == nil ||
+		c.Spec().Linux.Resources.CPU.Cpus == "" {
+		return fmt.Errorf("failed to find the container %q CPUs", c.ID())
+	}
+
+	content, err := ioutil.ReadFile(irqSmpAffinityFile)
+	if err != nil {
+		return err
+	}
+	currentIRQSMPSetting := strings.Trim(string(content), "\n")
+
+	newIRQSMPSetting, err := UpdateIRQSmpAffinityMask(c.Spec().Linux.Resources.CPU.Cpus, currentIRQSMPSetting, enable)
+	if err != nil {
+		return err
+	}
+	newIRQSMPSetting += "\n"
+	ioutil.WriteFile(irqSmpAffinityFile, []byte(newIRQSMPSetting), 0644)
+	return nil
+}
+
+func setCPUQuota(ctx context.Context, parentDir string, c *oci.Container, enable bool) error {
+	original := c.Spec().Linux.CgroupsPath
+	var rpath string
+	if strings.Contains(original, ":") {
+		// systemd managed, first replace the colon in "...slice:" with "/"
+		step1 := strings.Replace(original, ":", "/", 1)
+		// then replace colon in "crio:" with "-"
+		step2 := strings.Replace(step1, ":", "-", 1)
+		// lastly append ".scope"
+		rpath = step2 + ".scope"
```

Comment:
we could use https://godoc.org/github.com/opencontainers/runc/libcontainer/cgroups/systemd#ExpandSlice here

Reply:
@haircommander not sure if ExpandSlice will work. Here is the pattern of Spec().Linux.CgroupsPath, 
parentfolder:crio:containerid 
the real path is:
parentfolder/crio-containerid.slic

it doesn't look like ExpandSlice can do that conversion.

Reply:
I believe if you have a path:
parentfolder-subparentfolder:crio:containerid
it's going to be
/parentfolder/parentfolder-subparentfolder/crio/containerid.scope

so I believe this will fail without ExpandSlice. We're looking for the location of this cgroup on disk right?

Reply:
yes. a real example is to convert:
kubepods-pod344c6fac_64b4_431c_a449_10ed09a492ba.slice:crio:813c5bd769ca6b8198acf1e927236d10807630246702d94165e23a31515a6a1a

into an absolute path:
/sys/fs/cgroup/cpu/kubepods.slice/kubepods-pod344c6fac_64b4_431c_a449_10ed09a492ba.slice/crio-813c5bd769ca6b8198acf1e927236d10807630246702d94165e23a31515a6a1a.scope

I do not know why crio put the two colons around "crio". is it due to some convention? my code does do the job but if ExpandSlice can do the majic I would love to use it. but looking at its source: https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/systemd/common.go#L53
1. it expect a input string which .slice suffix
2. it expect "-" as separator
so as its doc says, its job is to convert  "test-a-b.slice" to  "/test.slice/test-a.slice/test-a-b.slice". This is not the input I'm observing. I'm testing it on openshift, with cgroup v1 layout.

Reply:
based on the input example provided above, ExpandSlice can take care of parent.slice/parent-subparent.slice portion, but not the crio-<containerid>.scope with the two colons around "crio". Let me know your thought on that

Reply:
ah yes, that seems more correct

also, if you look at `internal/config/cgmgr` there are some related functions (especially in systemd.go). perhaps we could utilize some code there?

Reply:
Given the input I'm observing on the parentDir and Spec().Linux.CgroupsPath, here is one way to use ExpandSlice:
ExpandSlice(parentDir) + "/" + "crio-" + fmt.Sprintf("%d", c.ID() + ".scope"

THis will at least take care if the parent folder; though the "crio...scope" portion still looks hardcoded.
what do you think ?